[Introduce] Continuous Worker mode in the Reindex From Snapshot process

jugal-chauhan opened the pull request:
### Description
RFS workers are currently short-lived. Each worker processes a single shard and then terminates.

While this model works well for ECS tasks, it causes unnecessary pod churn in EKS deployments, since Kubernetes marks pods that exit within 5 minutes as “failed” and places them in BackOff state.As a result, even successfully completed RFS workers were briefly treated as failures, introducing restart delays and slowing overall migration throughput.

This PR introduces continuous mode to enable long-running workers that can process multiple shards sequentially, while preserving existing ECS/test behavior. Changes introduced : 

1. New `--continuous-mode` for long running deployments
- CLI default: `false` (preserve non-continuous mode)
- Switched EKS deployment to use `true` for long living pods
- EKS pods should remain alive across shard boundaries to reduce restarts and operational overhead. Now in continuous mode, workers will acquire, process, re-acquire work indefinitely, stopping only when no migration work remains. 
- In single-run mode (ECS/tests), behavior remains unchanged where worker processes one shard and exits.

Previously, the worker’s main path did a single pass of work. That made continuous operation awkward and lease timers could leak across items with `--continuous-mode`. This PR introduced two layers:
- `runWorkItemLoop(...)` which owns the overall lifecycle loop for continuous mode and calls `processWorkItem(...)` once per work item attempt.
- `processWorkItem(...)` which owns the single-item attempt, including the per-item lease timer `LeaseExpireTrigger`, calls into existing `run(...)`, and returns a `boolean` telling the outer loop whether to continue.

2. Prevent work item double finalization during concurrent lifecycle events
- Long-lived workers increase the likelihood of race conditions between lease timeouts and shutdown hooks. This PR ensures clean transitions between work items and prevents concurrent finalization paths from interfering by
a. Adding `lastWorkItemFinalized` tracking to detect and skip a re-finalization
b. Scoping each work item to its own `LeaseExpireTrigger`

3. Simplifies CI/CD behavior by making process outcomes predictable across ECS, EKS, and test environments by defining consistent exit semantics:
- 0 = successful completion, 
- 1 = no work remaining, 
- 2 = lease timeout/processing failure
Replaces previous `NO_WORK_LEFT_EXIT_CODE = 3` for consistency.

4. New `ContinuousWorkerTest`
This new test explicitly validates contract to prevent regressions using 4 test cases:
- `No-work-left` exits with appropriate codes in both deployment modes
- `SIGTERM` during active shard processing triggers clean shutdown without double finalization
- Lease timeout resilience under network latency (ToxiProxy simulation) maintains process health in continuous mode
- Verify previous shard lease timers never fire during subsequent shard processing

### Issues Resolved
<!-- List any GitHub or Jira issues this PR will resolve -->

### Testing
<!-- Please provide details of testing done: unit testing, integration testing and manual testing -->

### Check List
- [ ] New functionality includes testing
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

AndreKurait commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -427,11 +430,15 @@ public static void main(String[] args) throws Exception {
             MDC.put(LOGGING_MDC_WORKER_ID, workerId); // I don't see a need to clean this up since we're in main
-            DocumentReindexer reindexer = new DocumentReindexer(targetClient,
+
+            // Target reindexer is safe to reuse across iterations
```
This comment can be removed

AndreKurait commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -448,37 +455,44 @@ public static void main(String[] args) throws Exception {
-        } catch (NoWorkLeftException e) {
-            log.atInfo().setMessage("No work left to acquire.  Exiting with error code to signal that.").log();
+            if (arguments.continuousMode) {
+                log.atInfo().setMessage("Running in continuous mode. RFS will process multiple work items until SIGTERM").log();
```
This log doesn't add much value and is inaccurate

AndreKurait commented on DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/LeaseExpirationTest.java:
```diff
@@ -217,7 +225,8 @@ private static int runProcessAgainstToxicTarget(
             "--initial-lease-duration", "PT20s",
-            "--source-version", sourceClusterVersion.getVersion().toString()
+            "--source-version", sourceClusterVersion.getVersion().toString(),
+            "--continuous-mode", "false"
```
Why does this test need to change at all if --continuous-mode is false?

jugal-chauhan replied:
reverted the change.

AndreKurait commented on DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/ContinuousWorkerTest.java:
```diff
@@ -0,0 +1,219 @@
+ *  - exits cleanly when interrupted (simulated SIGTERM)
+ */
+@Slf4j
+@Disabled("Test disabled from Gradle builds")
```
Is this going to be re-enabled before merging in?

AndreKurait replied:
I'd like to see logic from LeaseExpirationTest with tested with continuous mode on and off, either here or in the LeaseExpirationTest

AndreKurait commented on DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/ContinuousWorkerTest.java:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    @Test
+    void testContinuous_ES710_to_OS13() throws Exception {
```
Can you imagine any bug getting introduced that would occur by-version with this?

AndreKurait commented on DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/ContinuousWorkerTest.java:
```diff
@@ -0,0 +1,219 @@
+            var sourceOps = new ClusterOperations(sourceCluster);
+            var targetOps = new ClusterOperations(targetCluster);
+
+            String indexBody = "{\"settings\":{\"number_of_shards\":1,\"number_of_replicas\":0}}";
```
Could we have this with 2 shards to make sure it continues through both?

AndreKurait commented on DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/ContinuousWorkerTest.java:
```diff
@@ -0,0 +1,219 @@
+            "--source-version", sourceCluster.getContainerVersion().getVersion().toString(),
+            "--session-name", sessionName,
+            "--index-allowlist", index,
+            "--no-work-retry-delay-seconds", String.valueOf(noWorkDelaySeconds)
```
What is this argument?

AndreKurait commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -756,4 +771,85 @@ private static void confirmShardPrepIsComplete(
+            // Ensure nothing stale can be invoked later
+            cancellationRunnableRef.set(null);
+        }
+        return arguments.continuousMode && !Thread.currentThread().isInterrupted() && !cleanShutdownCompleted.get();
```
I think `!Thread.currentThread().isInterrupted()` is redundant here since you return false above

jugal-chauhan commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -756,4 +771,85 @@ private static void confirmShardPrepIsComplete(
+
+            if (completionStatus == CompletionStatus.NOTHING_DONE) {
+                if (arguments.continuousMode) {
+                    shortIdle();
```
shortidle() has been removed entirely

jugal-chauhan commented on DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/SourceTestBase.java:
```diff
@@ -5,14 +5,12 @@
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.Paths;
+import java.nio.file.*;
```
reverted the change

jugal-chauhan commented on DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/SourceTestBase.java:
```diff
@@ -355,15 +352,38 @@ public static CompletionStatus migrateDocumentsWithOneWorker(
     public static void deleteTree(Path path) throws IOException {
-        try (var walk = Files.walk(path)) {
-            walk.sorted(Comparator.reverseOrder()).forEach(p -> {
+        if (path == null) return;
```
Reverted this change for now

AndreKurait commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -448,43 +449,75 @@ public static void main(String[] args) throws Exception {
-            run(
-                new LuceneIndexReader.Factory(sourceResourceProvider),
-                reindexer,
+            // Unified path: always run at least once; continue if continuousMode=true.
```
remove comment

AndreKurait commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -448,43 +449,75 @@ public static void main(String[] args) throws Exception {
+                    unpackerFactory
+                );
+            } catch (IOException e) {
+                log.atError().setCause(e).setMessage("IO error during work item processing").log();
```
I'm not sure what this is adding beyond the caller's `Unexpected error running RfsWorker` when exceptions are thrown

AndreKurait commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -756,4 +799,80 @@ private static void confirmShardPrepIsComplete(
+            cancellationRunnableRef.set(null);
+        }
+        // loop continuation logic for continuous mode
+        return arguments.continuousMode && !cleanShutdownCompleted.get();
```
In what case would you expect this line to run and cleanShutdownCompleted == true?

jugal-chauhan replied:
This line will get executed when the RFS worker's iteration was normal (made some progress/completed work). 

This line executes when the `run(....)` method returned a status other than
- all above described exceptions like NoWorkLeftException or IO or InterruptedException
- CompletionStatus.NOTHING_DONE

It returns `true` to continue with the outer loop
- when continuous mode == true
- and cleanShutdownCompleted == false

It returns `false` to stop the loop
- when continuous mode == false
- or cleanShutdownCompleted == true

gregschohn commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -502,11 +562,22 @@ private static void executeCleanShutdownProcess(
             return;
         }
         var workItemAndDuration = workItemRef.get();
-        log.atInfo().setMessage("Marking progress: " + workItemAndDuration.getWorkItem().toString() + ", at doc " + progressCursor.get().getProgressCheckpointNum()).log();
```
did you get a warning for this line?  I had thought that with ErrorProne these were supposed to be build errors.  I'm wondering how it persisted

AndreKurait commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -449,48 +451,106 @@ public static void main(String[] args) throws Exception {
             cleanShutdownCompleted.set(true);
-            System.exit(NO_WORK_LEFT_EXIT_CODE);
+
+            if (arguments.continuousMode) {
```
Can you help me understand this branch? Both sides of this if run when `shouldContinue=false` and i have a hard time seeing how you would know this was succesful

AndreKurait commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -449,48 +451,106 @@ public static void main(String[] args) throws Exception {
-                reindexer,
-                progressCursor,
-                workCoordinator,
+            runWorkItemLoop(
```
Probably more straight forward to embed the do while here.

```
do {
} while(processWorkItem(
                initialLeaseDuration,
                continuousMode,
                snapshotName,
                previousSnapshotName,
                experimentalDeltaMode,
                indexAllowlist,
                maxShardSizeBytes,
                workItemRef,
                progressCursor,
                workCoordinator,
                workItemTimeProvider,
                cleanShutdownCompleted,
                lastWorkItemFinalized,
                context,
                cancellationRunnableRef,
                sourceResourceProvider,
                reindexer,
                unpackerFactory
 ));
```

AndreKurait commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -569,10 +643,19 @@ private static void exitOnLeaseTimeout(
         }
-        cleanShutdownCompleted.set(true);
-        System.exit(PROCESS_TIMED_OUT_EXIT_CODE);
+        // continuous mode: keep shutdown hook active for next item checkpointing
```
please revert

AndreKurait commented on DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/ContinuousWorkerTest.java:
```diff
@@ -0,0 +1,430 @@
+            SearchClusterContainer.ContainerVersion targetClusterVersion
+    ) {
+        final var testSnapshotContext = SnapshotTestContext.factory().noOtelTracking();
+        var tempDirSnapshot = Files.createTempDirectory("rfs_no_work_left_snap");
```
We should be able to use annotated temp directory in the class

gregschohn commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -387,33 +394,25 @@ public static void main(String[] args) throws Exception {
         var workItemTimeProvider = new WorkItemTimeProvider();
         var coordinatorFactory = new WorkCoordinatorFactory(targetVersion, arguments.indexNameSuffix);
         var cleanShutdownCompleted = new AtomicBoolean(false);
+        var lastWorkItemFinalized = new AtomicReference<String>();
```
Can you please add a comment describing how the value here transitions through the different execution paths and how this value is read to influence the execution?  That will help me understand this much better.

I'm also trying to figure out if you're trying to use this for concurrency control or just as a reference (the split get and set makes me thing just as a reference).

gregschohn commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -757,4 +839,93 @@ private static void confirmShardPrepIsComplete(
+                    return true;
+                } else {
+                    log.atInfo().setMessage("No progress made (non-continuous mode). Exiting with error code (no-work-left) to signal that.").log();
+                    System.exit(NO_WORK_LEFT_EXIT_CODE);
```
Generally try to exit ONLY at the top level and in as few places as possible.  Calling exit means that you won't be able to test that function.  This is another reason to flatten the logic more.

Notice that we DO call exit within the lease expiration handler - that's purposeful, since we can't reliably get back onto the main thread and into the main() method to do it there.

jugal-chauhan replied:
Fixed. All System.exit() calls have been moved to top-level control flow in main()

Lower level methods (processWorkItem, run, etc.) now return result enums instead of calling System.exit().

The only remaining direct exit is in exitOnLeaseTimeout(), where it’s required for reliability.

jugal-chauhan commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -72,8 +73,9 @@
+    public static final int SUCCESS_EXIT_CODE = 0;
+    public static final int NO_WORK_LEFT_EXIT_CODE = 1;
     public static final int PROCESS_TIMED_OUT_EXIT_CODE = 2;
-    public static final int NO_WORK_LEFT_EXIT_CODE = 3;
```
I have reverted this change to make this exit code remain 3

jugal-chauhan commented on DocumentsFromSnapshotMigration/src/main/java/org/opensearch/migrations/RfsMigrateDocuments.java:
```diff
@@ -502,11 +562,22 @@ private static void executeCleanShutdownProcess(
+
+        // Check if this work item was already finalized by exitOnLeaseTimeout
+        if (currentWorkItemId.equals(lastWorkItemFinalized.get())) {
+            log.atInfo().setMessage("Work item {} already finalized by lease timeout handler").addArgument(currentWorkItemId).log();
```
I added coverage via the lease timeout and SIGTERM tests in ContinuousWorkerTest, which simulates a clean shutdown and timeout finalization. The guard ensures a work item isn’t marked complete twice by both the lease handler and shutdown hook.

jugal-chauhan commented:
Update: I’ve removed `lastWorkItemFinalized` entirely.

The lease timeout path now sets a single, authoritative flag `cleanShutdownCompleted.set(finalized)` before `System.exit(PROCESS_TIMED_OUT_EXIT_CODE)`.

The shutdown hook early returns if:
	1.	`cleanShutdownCompleted.get()` is true, or
	2.	`workItemRef.get() == null || progressCursor.get() == null`

With those two guards, the shutdown hook won’t attempt to mark completion after a timeout path has already finalized (or when there’s nothing to finalize). This makes the extra “last finalized ID” state redundant, so I removed it.

jugal-chauhan commented:
After further discussions with the team, we've decided to proceed with the changes proposed in PR #1946 and PR #1939 instead. These two PRs will bring in the functionality originally intended by this PR in a more comprehensive manner.

Thank you for the review and feedback!
